frame/contracts/src/chain_extension.rs

```diff
@@ -0,0 +1,393 @@
+	///
+	/// This does **not** charge any weight. The caller must make sure that the an
+	/// appropriate amount of weight is charged **before** reading from contract memory.
+	/// The reason for that is that usually the costs for reading data and processing
+	/// said data cannot be separated in a benchmark. Therefore a chain extension would
+	/// charge the overall costs either using `max_len` (worst case approximation) or using
+	/// [`in_len()`](Self::in_len).
+	pub fn read(&self, max_len: u32) -> Result<Vec<u8>> {
+		self.inner.runtime.read_sandbox_memory(
+			self.inner.input_ptr,
+			self.inner.input_len.min(max_len),
+		)
+	}
+
+	/// Reads `min(buffer.len(), in_len) from contract memory.
+	///
+	/// This takes a mutable pointer to a buffer fills it with data and shrinks it to
+	/// the size of the actual data. Apart from supporting pre-allocated buffers it is
+	/// equivalent to to [`read()`](Self::read).
+	pub fn read_into(&self, buffer: &mut &mut [u8]) -> Result<()> {
+		let len = buffer.len();
+		let sliced = {
+			let buffer = core::mem::take(buffer);
+			&mut buffer[..len.min(self.inner.input_len as usize)]
+		};
+		self.inner.runtime.read_sandbox_memory_into_buf(
+			self.inner.input_ptr,
+			sliced,
+		)?;
+		*buffer = sliced;
+		Ok(())
+	}
+
+	/// Reads `in_len` from contract memory and scale decodes it.
+	///
+	/// This function is secure and recommended for all input types of fixed size
+	/// as long as the cost of reading the memory is included in the overall already charged
+	/// weight of the chain extension. This should usually be the case when fixed input types
+	/// are used. Non fixed size types (like everything using `Vec`) usually need to use
+	/// [`in_len()`](Self::in_len) in order to properly charge the necessary weight.
+	pub fn read_as<T: Decode>(&mut self) -> Result<T> {
+		self.inner.runtime.read_sandbox_memory_as(
+			self.inner.input_ptr,
+			self.inner.input_len,
+		)
+	}
+
+	/// The length of the input as passed in as `input_len`.
+	///
+	/// A chain extension would use this value to calculate the dynamic part of its
+	/// weight. For example a chain extension that calculates the hash of some passed in
+	/// bytes would use `in_len` to charge the costs of hashing that amount of bytes.
+	/// This also subsumes the act of copying those bytes as a benchmarks measures both.
+	pub fn in_len(&self) -> u32 {
+		self.inner.input_len
+	}
+}
+
+/// Functions to use the output arguments as pointer to a buffer.
+impl<'a, 'b, E: Ext, S: state::BufOut> Environment<'a, 'b, E, S>
+where
+	<E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+{
+	/// Write the supplied buffer to contract memory.
+	///
+	/// If the contract supplied buffer is smaller than the passed `buffer` an `Err` is returned.
+	/// If `allow_skip` is set to true the contract is allowed to skip the copying of the buffer
+	/// by supplying the guard value of [`u32::max_value()`] as `input_len_ptr`. The
```

Comment:
I'm not sure I follow, isn't `output_ptr` instead of `input_len_ptr` which must hold u32::max_value to be skippable ?
```suggestion
	/// by supplying the guard value of [`u32::max_value()`] as `output_ptr`. The
```

Reply:
Yes this was an oversight. There is no input ptr when writing data. I changed the docs.